src/sys/unix/socket.rs

```diff
@@ -0,0 +1,274 @@
+#[cfg(any(feature = "tcp", feature = "udp"))]
+use crate::sys::unix::net::from_socket_addr;
+#[cfg(feature = "tcp")]
+use crate::sys::unix::net::to_socket_addr;
+use std::io::Result;
+#[cfg(any(
+    feature = "tcp",
+    all(any(target_os = "ios", target_os = "macos"), feature = "udp")
+))]
+use std::mem;
+#[cfg(feature = "tcp")]
+use std::mem::MaybeUninit;
+#[cfg(any(feature = "tcp", feature = "udp"))]
+use std::net::SocketAddr;
+use std::os::unix::io::{AsRawFd, FromRawFd, IntoRawFd, RawFd};
+
+#[derive(Debug)]
+pub(crate) struct Socket {
+    fd: libc::c_int,
+}
+
+impl Socket {
+    pub(crate) fn new(
+        domain: libc::c_int,
+        socket_type: libc::c_int,
+        protocol: libc::c_int,
+    ) -> Result<Self> {
+        #[cfg(any(
+            target_os = "android",
+            target_os = "dragonfly",
+            target_os = "freebsd",
+            target_os = "linux",
+            target_os = "netbsd",
+            target_os = "openbsd"
+        ))]
+        let socket_type = socket_type | libc::SOCK_NONBLOCK | libc::SOCK_CLOEXEC;
+
+        let socket = syscall!(socket(domain, socket_type, protocol))?;
+
+        // Darwin and Solaris do not have SOCK_NONBLOCK or SOCK_CLOEXEC.
+        //
+        // In order to set those flags, additional `fcntl` sys calls must be
+        // performed. If a `fnctl` fails after the socket has been created,
+        // `close` ensures the socket does not leak.
+        #[cfg(any(target_os = "ios", target_os = "macos", target_os = "solaris"))]
+        syscall!(fcntl(socket, libc::F_SETFL, libc::O_NONBLOCK))
```

Comment:
This closure maybe better suited as a method that sets an arbitrary fd to be nonblocking and cloexec etc., so the same logic will be easily applicable after using from_raw_fd on an arbitrary FD (maybe also F_GETFL the original flags before setting).

Reply:
> so the same logic will be easily applicable after using from_raw_fd on an arbitrary FD

Yes that could be useful! For the targets where this currently runs, we know these flags are not set by default and the FD was just created, so an extra `F_GETFL` would probably be unnecessary.

For this PR however I was trying to keep it as similar as possible without changing too much, since the intent was to just introduce the `Socket` type. I'm not sure if this PR will land due to discussion above, but if it makes it in in the future for supporting other kinds of sockets it would be something to discuss!

Reply:
Edit: Ah sorry discussion briefly took place in the Tokio discord and not on this PR. Maybe @Thomasdezeeuw can echo some of their thoughts on the PR as well. Essentially, we still aren't sure if a `Socket` type (private or not) still makes sense in Mio.